### PR TITLE
feat(cdnurl): Add CDNJS support

### LINKED ISSUE #5253: New package manager: cdnjs
**What would you like Renovate to be able to do?**

Scan files for CDNJS URLs and then "renovate" them if necessary.

**Describe the solution you'd like**

Default to scanning .html files and let users extend it into others via `fileMatch`.

We hopefully don't need to be "html aware" and instead just look for any fully-matching CDNJS url and extract it.

I am aware of this format for URLs we can support first:

- `https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.15/lodash.core.js`

But I wonder if there are any alternatives or legacy URL formats we should be aware of too. This is also commonly used:

- `//cdnjs.cloudflare.com/ajax/libs/simple-line-icons/2.4.1/css/simple-line-icons.min.css` (i.e. no `https:` prefix)

**Describe alternatives you've considered**

Such URLs get stale otherwise.

**Additional context**

https://blog.cloudflare.com/javascript-libraries-are-almost-never-updated/

### PR DESCRIPTION
**Pull request by zharinov**
Examples: 

https://github.com/zharinov/renovate-test-cdnjs/pulls
https://github.com/zharinov/gifSearchEngineWithBabelPolyfill/pulls

Closes #5253

### CODE REVIEW

File: lib/config/definitions.ts [hunk 1]
```diff
@@ -1557,6 +1557,18 @@ const options: RenovateOptions[] = [
...
   },
+  {
+    name: 'cdnjs',
+    description: 'Configuration object for CDNJS assets',
+    stage: 'repository',
+    type: 'object',
+    default: {
+      enabled: true,
+      fileMatch: ['\\.html?$'],
+      versionScheme: VERSION_SCHEME_NPM,
```
Comment (rarkins) [inline lib/config/definitions.ts]: Maybe semver instead, assuming it's not possible to define ranges.

File: lib/manager/cdnjs/extract.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    offset += match.index + wholeSubstr.length;
+    rest = content.slice(offset);
+    match = rest.match(regex);
+
+    deps.push({
+      datasource: DATASOURCE_CDNJS,
+      depName,
+      lookupName: `${depName}/${asset}`,
+      currentValue,
+      fileReplacePosition,
```
Comment (rarkins) [inline lib/manager/cdnjs/extract.ts]: Shouldn't this be under `managerData`?

File: lib/manager/cdnjs/update.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  const versionPart = rightPart.slice(0, versionClosePosition);
+  const version = versionPart.trim();
+  if (version === newValue) {
+    return fileContent;
+  }
+  if (version === currentValue) {
+    const replacedPart = versionPart.replace(version, newValue);
+    return leftPart + replacedPart + restPart;
+  }
+  logger.debug({ depName, version, currentValue, newValue }, 'Unknown value');
```
Comment (rarkins) [inline lib/manager/cdnjs/update.ts]: This path will be taken if there are any modifications to the html prior to the CDNJS links, right?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/cdnjs/update.ts (hunk 3 above)
Comment (zharinov) [reply lib/manager/cdnjs/update.ts]: It returns `null` if content was changed in the way different than just change `currentValue` to `newValue`

File: lib/datasource/index.ts [hunk 4]
```diff
@@ -38,7 +38,7 @@ export * from './common';
 
 const datasources: Record<string, Datasource> = {
   cargo,
-  cdnjs,
+  cdn: cdnjs,
```
Comment (rarkins) [inline lib/datasource/index.ts]:
```suggestion
  cdnjs,
```

File: lib/manager/cdn/extract.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { PackageFile, PackageDependency } from '../common';
+import { DATASOURCE_CDNJS } from '../../constants/data-binary-source';
+
+export function extractPackageFile(content: string): PackageFile {
+  const deps: PackageDependency[] = [];
+
+  const regex = /\/\/cdnjs\.cloudflare\.com\/ajax\/libs\/(?<depName>[^/]+?)\/(?<currentValue>[^/]+?)\/(?<asset>[-_.a-zA-Z0-9]+)/;
```
Comment (JamieMagee) [inline lib/manager/cdn/extract.ts]: This is fine for an initial implementation, but we should probably be parsing HTML and only updating `<script>` tags.

File: lib/manager/cdn/update.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  const { fileReplacePosition } = managerData;
+  const leftPart = fileContent.slice(0, fileReplacePosition);
+  const rightPart = fileContent.slice(fileReplacePosition);
+  const versionClosePosition = rightPart.indexOf('/');
+  const restPart = rightPart.slice(versionClosePosition);
+  const versionPart = rightPart.slice(0, versionClosePosition);
+  const version = versionPart.trim();
+  if (version === newValue) {
+    return fileContent;
+  }
```
Comment (JamieMagee) [inline lib/manager/cdn/update.ts]: What's the possible scenario where this could occur?

File: lib/manager/cdn/extract.ts (hunk 5 above)
Comment (JamieMagee) [reply lib/manager/cdn/extract.ts]: I just had another though. If any sites are using [Subresource integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) this implementation will break the site.

It appears that the CDNJS API doesn't provide file hashes for hosted dependencies.

Comment (JamieMagee) [reply lib/manager/cdn/extract.ts]: See cdnjs/new-website#246

File: test/manager/cdn/_fixtures/sample.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+      <script type="text/javascript"
```
Comment (JamieMagee) [inline test/manager/cdn/_fixtures/sample.html]: Some other suggestions for test cases:
1. relative protocol i.e. `//cdnjs.cloudflare.com`
2. HTTP sample
3. An invalid script tag i.e. `<script src="..." />`. This is invalid, but browsers will accept it.
4. CSS example for cdnjs

**Comment by JamieMagee** [comment]
I can't sign off on this, due to my comment around SRI. I think supporting CDNJS has the possibility to have a wide ranging appeal, and be very popular.

That being said we should be careful to not break any existing sites, as this would definitely turn people off of Renovate, and dependency updates in general.

**Comment by zharinov** [comment]
Just to summarise:
- HTML
  - [ ] Search cdn url inside `<script>` tag
  - [ ] Skip update for tags containing `script.integrity !== 'undefined'`, if we have no possibility to update it accordingly
- Or, in more generic way
  - [x] Search and update URLs inside file with any format (e.g. [gruntfiles](https://github.com/sashberd/homework/blob/e653a14e197cd83eec1a3bfb559bda8ec03e7fa3/edtwebapi/js/grunt/Gruntfile.js) or [Python flask framework](https://github.com/tkalus-old/fogspoon/blob/e0ec46e2f8a3ab87cc526964df206a4cc482cc60/fogspoon/frontend/assets.py))
  - [ ]  Search for SRI hashes inside (same?) file and update them (or fail?)

**Comment by rarkins** [comment]
On one hand:

- Great catch, @JamieMagee. In cases where a subresource integrity hash is there, the current approach would break things

On the other hand:

- I bet there are many locations, including non-html, where regular URLs exist and would benefit from this

I'm trying to decide between two approaches:
1. Rename this manager `cdnurl` (a more accurate name), give it an empty `fileMatch` default, and add small documentation about how users must opt-in
2. Rename this manager `html`, parse the html files looking for script includes, and:
 - Skip resource integrity ones in phase 1
 - Add integrity support in phase 2

**Comment by zharinov** [comment]
If we're about to also add UNPKG datasource, it supports [hashes retrieval from api](https://unpkg.com/react@16.12.0/umd/react.production.min.js?meta).

**Comment by rarkins** [comment]
Those two above approaches aren't mutually exclusive, so @zharinov what do you think about making option 1 changes first, and then starting on option 2 phase 1?

Re: hashes, we can retrieve the file in question and calculate its hash, presumably. We do the same for some `bazel` dependencies already - it's done as part of the update function and not artifacts.

If cdnjs doesn't have an easy way to get hashes, I doubt many people would use them with cdnjs.

File: lib/manager/cdn/update.ts (hunk 6 above)
Comment (rarkins) [reply lib/manager/cdn/update.ts]: Should happen whenever the file has been updated already in a previous run

**Pull request merged**